sumomongodbatlascollector/api.py

```diff
@@ -51,66 +53,65 @@ class FetchMixin(MongoDBAPI):
 
     def fetch(self):
         log_type = self.get_key()
-        output_handler = OutputHandlerFactory.get_handler(self.collection_config['OUTPUT_HANDLER'], path=self.pathname, config=self.config)
-        url, kwargs = self.build_fetch_params()
-        self.log.info(f'''Fetching LogType: {log_type} kwargs: {kwargs}''')
-        state = None
-        payload = []
+        output_handler = None
         try:
-
+            output_handler = OutputHandlerFactory.get_handler(self.collection_config['OUTPUT_HANDLER'], path=self.pathname, config=self.config)
+            url, kwargs = self.build_fetch_params()
+            self.log.info(f'''Fetching LogType: {log_type} kwargs: {kwargs}''')
+            state = None
             fetch_success, content = ClientMixin.make_request(url, method="get", logger=self.log, TIMEOUT=self.collection_config['TIMEOUT'], MAX_RETRY=self.collection_config['MAX_RETRY'], BACKOFF_FACTOR=self.collection_config['BACKOFF_FACTOR'], **kwargs)
-            if fetch_success and len(content) > 0:
+            if fetch_success and content:
                 payload, state = self.transform_data(content)
-                #Todo Make this atomic if after sending -> Ctrl - C happens then it fails to save state
                 params = self.build_send_params()
                 send_success = output_handler.send(payload, **params)
                 if send_success:
                     self.save_state(**state)
                     self.log.info(f'''Successfully sent LogType: {self.get_key()} Data: {len(content)}''')
                 else:
                     self.log.error(f'''Failed to send LogType: {self.get_key()}''')
-            elif fetch_success and len(content) == 0:
+            elif fetch_success:
                 self.log.info(f'''No results window LogType: {log_type} kwargs: {kwargs} status: {fetch_success}''')
                 is_move_fetch_window, new_state = self.check_move_fetch_window(kwargs)
                 if is_move_fetch_window:
                     self.save_state(**new_state)
                     self.log.debug(f'''Moving fetched window newstate: {new_state}''')
             else:
                 self.log.error(f'''Error LogType: {log_type} status: {fetch_success} reason: {content}''')
+        except Exception as e:
+            self.log.exception(f'''Exception occurred while fetching LogType: {log_type}''')
         finally:
-            output_handler.close()
-            self.log.info(f'''Completed LogType: {log_type} curstate: {state} datasent: {len(payload)}''')
-
+            if output_handler:
+                output_handler.close()
+            self.log.info(f'''Completed LogType: {log_type} curstate: {state} datasent: {len(payload) if 'payload' in locals() else 0}''')
```

Comment:
earlier empty payload was there lets have that payload=[] before try,  normally no logic should be present inside formatted strings it makes it unreadable.